core/unix/os.c

```diff
@@ -2751,14 +2751,14 @@ static void
 replace_thread_id(thread_id_t old, thread_id_t new)
 {
 #ifdef HAVE_TLS
-    int32_t new_tid = (int32_t) new; /* can't use thread_id_t since it's 64-bits on x64 */
+    ptr_int_t new_tid = new;
     ASSERT(is_thread_tls_initialized());
     DOCHECK(1, {
-        int32_t old_tid; /* can't use thread_id_t since it's 64-bits on x64 */
-        READ_TLS_INT_SLOT_IMM(TLS_THREAD_ID_OFFSET, old_tid);
+        ptr_int_t old_tid;
+        READ_TLS_SLOT_IMM(TLS_THREAD_ID_OFFSET, old_tid);
```

Comment:
This will read the adjacent field (or maybe padding, did not check the struct): it would have to be truncated (or have a custom READ_TLS_SLOT_TIDSZ_IMM as suggested in original comment, or have series of ifdefs or sizeof checks here).